scripts/build_test.py

```diff
@@ -14,29 +14,270 @@
+                source_dir_file_count, target_dir_file_count) in
+            incorrectFileCount.exception)
+
+    def test_match_filename_with_hashes(self):
+        """Test _match_filename_with_hashes to raise exception:
+            1) When there is an empty hash dict.
+            2) When a filename is expected to contain hash but does not.
+            3) When there is a hash in filename that cannot be found in
+                hash dict.
+        """
+        # Final filepath example: base.240933e7564bd72a4dde42ee23260c5f.html.
+        file_hashes = dict()
+        base_filename = 'base.html'
+        with self.assertRaises(ValueError) as emptyHashDict:
+            build._match_filename_with_hashes(base_filename, file_hashes)
+        self.assertTrue('Hash dict is empty' in emptyHashDict.exception)
+
+        file_hashes = {base_filename: random.getrandbits(128)}
+        bad_filepath = 'README'
+        with self.assertRaises(ValueError) as filepathWithoutExtension:
+            build._match_filename_with_hashes(bad_filepath, file_hashes)
+        # Generate a random hash dict for base.html.
+        self.assertTrue(
+            'Filepath has less than 2 partitions after splitting'
+            in filepathWithoutExtension.exception)
+
+        with self.assertRaises(ValueError) as noHashInFilename:
+            build._match_filename_with_hashes(base_filename, file_hashes)
+        # Generate a random hash dict for base.html.
+        self.assertTrue(
+            '%s is expected to contain MD5 hash' % base_filename
+            in noHashInFilename.exception)
+
+        hashed_base_filename = build._insert_hash(
+            base_filename, random.getrandbits(128))
+        with self.assertRaises(KeyError) as incorrectHashInFilename:
+            build._match_filename_with_hashes(hashed_base_filename, file_hashes)
+        self.assertTrue(
+            'Hash from file named %s does not match hash dict values'
+            % hashed_base_filename in incorrectHashInFilename.exception)
+
+    def test_process_html(self):
+        """Test process_html to remove whitespaces and hash filepaths."""
+        base_source_path = os.path.join(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'], 'pages',
+            'base.html')
+        # Prepare a file_stream object from StringIO.
+        minified_html_file_stream = StringIO.StringIO()
+        # Obtain actual file hashes of /templates to add hash to all filepaths
+        # within the HTML file. The end result will look like:
+        # E.g <script ... app.js></script>
+        # --> <script ... app.[hash].js></script>.
+        file_hashes = build.get_file_hashes(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'])
+        # Reading from actual base.html file and not simply a dummy file.
+        build._ensure_files_exist([base_source_path])
+        # Assert that base.html has white spaces.
+
+        with open(base_source_path, 'r') as source_base_file:
```

Comment:
There's way too much logic in this test. Structure your test base.html file such that you can just do a bunch of simple equality comparisons here. Having lots of "if" statements and logic in tests means that the tests become just as error-prone as the code.

Reply:
No 'if' statement in test_process_html!